Audit fixes & Unit Tests

kevinbeauregard opened the pull request:
The first commit adds changes from auditors:
- Contract changes
- Unit Tests

Subsequent commits include:
- bringing dependencies up to date
- cleaning up auditor code (lint / style)
- updating test assertions to better make use of Chai & OZ helpers
- Notable Change: Separate 'Withdraw All' logic into its own functions (ETH + ERC20) to avoid using '-1' as a switch case

kevinbeauregard committed:
Add tests and code changes from audit

kevinbeauregard committed:
Update NPM dependencies

kevinbeauregard committed:
Solhint fixes

kevinbeauregard committed:
Separate Withdraw -> Withdraw/WithdrawAll; get tests passing

jgarzik commented on contracts/Acct.sol:
```diff
@@ -35,72 +35,112 @@ contract Acct is Ownable, IAcct {
+        transferOwnershipToNFT(REGISTRY_ADDR);
+    }
+
+    function transferOwnershipToNFT(address registry) public onlyOwner {
```
Lack of the `override` keyword on an external function indicates that `transferOwnershipToNFT` should be added to the interface file.

kevinbeauregard replied:
That seems correct. This came from the auditors - I'll review their notes and then make the appropriate changes

jgarzik commented on contracts/Acct.sol:
```diff
@@ -35,72 +35,112 @@ contract Acct is Ownable, IAcct {
+        require(amount > 0, "Amount must be greater than zero");
         // ascertain available balance
-        assetBalance = ERC20(_assetAddress).balanceOf(address(this));
+        uint256 assetBalance = ERC20(_assetAddress).balanceOf(address(this));
```
The elimination of the -1 path implies that the withdraw-all case now calls `balanceOf()` repeatedly in the same code execution path.

kevinbeauregard replied:
True - however, the downstream check (L101 or L141) isn't actually necessary as in both cases the transfer will fail if the amount exceeds the balance. I will remove the downstream checks.

jgarzik reviewed (changes requested):
1. Left some comments relating to minor changes.
2. Is it possible for an "evil Acct", not from our factory, to be used with this system?

kevinbeauregard committed:
Remove duplicate TransferOwnership func & balance calls

kevinbeauregard commented:
@jgarzik changes made to address minor points
- removed duplicate transfer function and hardcoded registry address
- removed duplicate balance checks as extraneous - only negative side effect of this is a less pretty error message if you try to withdraw more Eth than the contract has, which is not an issue IMO

Regarding point 2 above, I dont *think* so, but it is possible to transfer ownership of a legitimate Acct to a malicious registry. I'll address that in a subsequent PR. We also need more test coverage.

jgarzik commented:
@kevinbeauregard an Acct can be owned by anyone or anything.
